JVM-2340: java.lang.IllegalArgumentException: wrong number of argumen…

navyxliu opened the pull request:
…ts: 0 expected: 1

This patch refactors PEAState::put_field. After we iterate all fields of the current object in materialization, we need to check all objects in AllocationState.  If the state of an object is virtual, we check its fields. It's possible that a virtual object has a field which is equals to the object we are materializing. Despite the object is still virtual, we may have committed field 'var' to memory via the prior putfield.  we don't have a way to revert, but we can emit a new store to overwrite it. The new store post-dominates the older one, they should coalesce.  

A new regression test 'MaterializeMember' is this case. We reduce it from TestNG. This patch can fixed the remaining failures of jdk:tier1 and langtools:tier1.

Commit:
JVM-2340: java.lang.IllegalArgumentException: wrong number of arguments: 0 expected: 1

Commit:
fix ctw

caojoshua commented on src/hotspot/share/opto/parseHelper.cpp:
```diff
@@ -617,67 +667,51 @@ Node* PEAState::materialize(GraphKit* kit, Node* var) {
-          field_type = TypeOopPtr::make_from_klass(field->type()->as_klass());
-        } else {
-          field_type = Type::BOTTOM;
+    // back from DFS, we still need to check again for all virtual states.
```
I'm ok with this for now so we can pass regression tests. In the future, I think we will need to revisit our PEA design. Iterating over every object every time we materialize can maybe be expensive.

navyxliu replied:
yes, we may have better data structure for this, but I think current move is acceptable.   We only iterate all objects of the current allocation state. It only tracks live objects of current basic block. 

Actually, I found that Graal has a similar look backward for each materialization. 
https://github.com/oracle/graal/blob/87e0f80b1b3c1621e1e0c9f4a3b0f27816b0142f/compiler/src/jdk.internal.vm.compiler/src/org/graalvm/compiler/virtual/phases/ea/PartialEscapeClosure.java#L554

Commit:
Change field_iterator from STL-style to Java-style.

I don't think it's a good idea to provide STL-style iterator. We don't
use it in <algorithm>. The price we may include:

1. operator==/operator!= overload.
2. operator-> overload
3. the sentinel object.

it's so unncessary. Java-style iterator is more concise.
